bin/verdicts.py

```diff
@@ -0,0 +1,496 @@
+        # testgroup -> testcase iterator
+        self._unknowns = {node: self.unknowns_iterator(node) for node in testgroups}
+        # testgroup -> testcase | None
+        self.first_unknown: dict[str, str | None] = {
+            node: next(self._unknowns[node]) for node in testgroups
+        }
+
+    # Allow `with self` to lock.
+    def __enter__(self):
+        self.lock.__enter__()
+
+    def __exit__(self, *args):
+        self.lock.__exit__(*args)
+
+    def unknowns_iterator(self, node):
+        """Yield the node's (yet) unknown children in lexicographic order."""
+        with self:
+            for child in self.children[node]:
+                if self._verdict[child] is not None:
+                    continue
+                yield child
+
+    def is_testgroup(self, node) -> bool:
+        """Is the given testnode name a testgroup (rather than a testcase)?
+        This assumes nonempty testgroups.
+        """
+        return node in self.children
+
+    def is_testcase(self, node) -> bool:
+        """Is the given testnode name a testcase (rather than a testgroup)?
+        This assumes nonempty testgroups.
+        """
+        return node not in self.children
+
+    def set(self, testcase, verdict: str | Verdict, duration: float | None):
+        """Set the verdict and duration of the given testcase (implying possibly others)
+
+        verdict can be given as a Verdict or as a string using either long or
+        short form ('ACCEPTED', 'AC', or Verdict.ACCEPTED).
+        """
+        with self:
+            if isinstance(verdict, str):
+                verdict = from_string(verdict)
+            self.duration[testcase] = duration
+            self._set_verdict_for_node(testcase, verdict)
+
+    def __getitem__(self, testnode) -> Verdict | None:
+        with self:
+            return self._verdict[testnode]
+
+    def salient_testcase(self) -> str:
+        """The testcase most salient to the root verdict.
+        If self['.'] is Verdict.ACCEPTED, then this is the slowest testcase.
+        Otherwise, it is the lexicographically first testcase that was rejected."""
+        with self:
+            match self['.']:
+                case None:
+                    raise ValueError("Salient testcase called before submission verdict determined")
+                case Verdict.ACCEPTED:
+                    # This implicitly assumes there is at least one testcase.
+                    return max((d, tc) for tc, d in self.duration.items() if d is not None)[1]
+                case _:
+                    return min(
+                        tc
+                        for tc, v in self._verdict.items()
+                        if self.is_testcase(tc) and v != Verdict.ACCEPTED
+                    )
+
+    def slowest_testcase(self) -> str:
+        """The slowest testcase."""
+        with self:
+            return max((v, tc) for tc, v in self.duration.items() if v is not None)[1]
```

Comment:
Why only return the name of the slowest test case? At the (only) use site, the duration is needed anyway :slightly_smiling_face: 

(Same for `salient_testcase`)

Reply:
Because this is cleaner, in that it's clear that the shown duration actually corresponds to the testcase instead of being some other duration.